Articles/Blog/WhenIsBetterToWriteAutomatedTests-src.md

```diff
@@ -0,0 +1,110 @@
+
+After reading the book "Working Effectively with Legacy Code", I resolved to turn over a new leaf and become a born-again Agile software engineer that would use TDD to write all new code and religiously apply the Legacy Code Change Algorithm to modify all legacy code (i.e. code without tests).
+Over the years since my conversion to an Agile software engineer, I often rigorously wrote automated tests for nearly every piece of software I touched.
+There where times  under schedule pressure that I slipped and reverted back to my old ways and failed to write sufficient automated tests and came to regret it.
+However, there were also times where my overzealous drive to religiously test everything (which usually came after the guilt of a recent slip of not testing enough) where I wasted time writing automated tests that were not worth the effort.
+And now after having had followed this road from 2007 until now having written and maintained thousands of automated tests over that time, I have come to realize that there are times when it is actually better not to write automated tests for some pieces of software!
+
+## Criteria for when it is better to not add automated tests
+
+I have come to realize that it often does not pay off to write automated tests for a piece of software when the following five things are true about that software:
+
+* **The damage done by a defect is minor.**
+* **The majority of defects will be obvious.**
+* **Fixing the majority of defects will be easy.**
+* **Manually testing the software after a change is easy.**
+* **Writing automated tests is hard or the tests will be hard to maintain.**
+
+If all five of the above criteria are satisfied, it is often better not to write automated tests for a piece of software.
+The time spent writing, maintaining, and running the automated tests does not have enough benefit to outweigh the cost.
+
+For example, some bash scripts used locally that load some modules and then configure, build, and run tests for some software may not be worth writing an automated test suite for because:
+
+* The damage done by not having everything complete correctly will be minor because you can just run it again.
+* A failure will be obvious because the software will not even build and no tests will be run.
+* The scripts do pretty basic stuff and will be easy to fix if they fail.
+* It would be difficult to write mock up the various commands run by the script.
+
+Therefore, it is often not worth the investment to write automated tests for locally run scripts of this type.
+
+## Example where automated tests should have been added
+
+However, if such scripts are being used to automate some important process like deploying the software to users where not performing the task correctly would cause non-trivial harm, then one should write some automated tests to protect the critical functionality of the scripts.
```

Comment:
> However, if such scripts are
A critical point worth remembering is that, if such scripts...

Reply:
Added this in https://github.com/betterscientificsoftware/betterscientificsoftware.github.io/pull/377/files#diff-ff08f44603f4e10f3095baf6cea0da6eR59.